### PR TITLE
Fixed #26682 -- Added support for Oracle identity columns.

### PR DESCRIPTION
**Pull request by felixxm**
Ticket [26682](https://code.djangoproject.com/ticket/26682). This PR contains few improvements on Oracle backend i.e.:
- autofield introspection,
- support for Oracle identity columns,
- support for use of Oracle sequences in bulk create.

### CODE REVIEW

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: django/db/backends/oracle/operations.py [hunk 1]
```diff
@@ -512,29 +491,44 @@ def combine_expression(self, connector, sub_expressions):
...
+                    placeholder = BulkInsertMapper.types.get(internal_type, '%s') % placeholder
+                # Add columns aliases to the first select to avoid "ORA-00918:
+                # column ambiguously defined" when two or more columns in the
+                # first select has the same values.
+                if not query:
+                    placeholder = "%s col_%s" % (placeholder, i)
+                select.append(placeholder)
             query.append('SELECT %s FROM DUAL' % ', '.join(select))
-        return ' UNION ALL '.join(query)
+        return 'SELECT * FROM (%s)' % ' UNION ALL '.join(query)
```
Comment (felixxm) [inline django/db/backends/oracle/operations.py]: If we use Oracle identity columns in tables in which we try to bulk insert then Oracle internally adds `sequence.nextval` to it. On the other hand we cannot use `sequence.nextval` in the following construct: _"A SELECT statement that is combined with another SELECT statement with the UNION, INTERSECT, or MINUS set operator"_ (see [documentation](http://docs.oracle.com/database/121/SQLRF/pseudocolumns002.htm#SQLRF00253)), hence I changed statement e.g.:
```sql
SELECT 'foo', 'foo', 'foo_sth' FROM dual
UNION ALL
SELECT 'bar', 'bar', 'bar_sth' FROM dual
```
to 
```sql
SELECT * FROM
  (SELECT 'foo' col_0, 'foo' col_1, 'foo_sth' col_2 FROM dual
   UNION ALL
   SELECT 'bar', 'bar', 'bar_sth' FROM dual)
```
I also added columns aliases to the first select to avoid `ORA-00918: column ambiguously defined` when two or more columns in the first select has the same values.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
I'm going to add release notes and documentation in the next few days.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
`schema.tests.SchemaTests.test_creation_with_db_table_double_quotes` failure is not related with this PR, it is related to Oracle test user and test tablespace changes in Jenkins settings. If we use e.g. `X_PULLREQUESTSORACLE_35` instead of `D4_PULLREQUESTSORACLE_35` everything works fine. It looks like some strip / quote bug in the Oracle backend. I will create separate ticket and PR for this.

**Comment by felixxm** [comment]
I created separate ticket [27843](https://code.djangoproject.com/ticket/27843)  and PR https://github.com/django/django/pull/8066 with `test_creation_with_db_table_double_quotes` fix.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
I rebased from master.

**Comment by felixxm** [comment]
I added release notes.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: tests/schema/tests.py [hunk 2]
```diff
@@ -1024,6 +1024,8 @@ def test_alter_implicit_id_to_explicit(self):
         Author.objects.create(name='Foo')
         Author.objects.create(name='Bar')
 
+    @unittest.skipIf(connection.vendor == 'oracle',
+                     "Modifying to the identity column isn't supported in Oracle")
```
Comment (shaib) [inline tests/schema/tests.py]: I don't think this used to be the case -- if it is, then:
A) The error should be caught and reported properly
B) A test should be added for proper error handling
C) It needs to be documented as a backwards-incompatible change -- it may break existing sets of migrations

**Comment by felixxm** [comment]
I added schema tests for changing the identity column to numeric and non-numeric type i.e.
- `schema.tests.SchemaTests.test_alter_auto_field_to_integer_field`,
- `schema.tests.SchemaTests.test_alter_auto_field_to_char_field`.

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: tests/schema/tests.py (hunk 2 above)
Comment (felixxm) [inline tests/schema/tests.py]: I added simple workaround (see [oracle/schema.py](https://github.com/django/django/pull/8055/files#diff-09834dce199a01ca699094cb994d2e49R71)).

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: django/db/backends/oracle/operations.py (hunk 1 above)
Comment (shaib) [inline django/db/backends/oracle/operations.py]: :+1: you may want to add part of that comment to the code

File: django/db/backends/oracle/schema.py [hunk 3]
```diff
@@ -56,22 +63,35 @@ def alter_field(self, model, old_field, new_field, strict=False):
...
+          nullable,
+        - update the table to transfer values from old to new,
+        - drop old column,
+        - rename the new column and possibly drop the nullable property.
         """
         # Make a new field that's like the new one but with a temporary
         # column name.
         new_temp_field = copy.deepcopy(new_field)
-        new_temp_field.null = True
+        new_temp_field.null = False if new_field.get_internal_type() == 'AutoField' else True
```
Comment (shaib) [inline django/db/backends/oracle/schema.py]: ```new_temp_field.null = (new_field.get_internal_type() != 'AutoField')```

File: django/db/backends/oracle/schema.py [hunk 4]
```diff
@@ -119,3 +139,21 @@ def _generate_temp_name(self, for_name):
...
+    def _is_identity_column(self, table_name, column_name):
+        with self.connection.cursor() as cursor:
+            cursor.execute("""
+                SELECT
+                    CASE WHEN identity_column = 'YES' THEN 1 ELSE 0 END
+                FROM user_tab_cols
+                WHERE table_name = %s AND
+                      column_name = %s
+            """, [self.normalize_name(table_name), self.normalize_name(column_name)])
+            rows = cursor.fetchone()
```
Comment (shaib) [inline django/db/backends/oracle/schema.py]: rename `rows` => `row`

File: docs/releases/2.0.txt [hunk 5]
```diff
@@ -120,7 +120,11 @@ CSRF
 Database backends
 ~~~~~~~~~~~~~~~~~
 
-* ...
+* Added support for the :class:`~django.db.models.AutoField`,
+  :class:`~django.db.models.BigAutoField` as a Oracle Identity Columns
```
Comment (shaib) [inline docs/releases/2.0.txt]: :class:`~django.db.models.BigAutoField` fields as Oracle Identity Columns

File: docs/releases/2.0.txt [hunk 6]
```diff
@@ -120,7 +120,11 @@ CSRF
...
+  (:ticket:`26682`).
+
+* Oracle database introspection now works with auto fields.
```
Comment (shaib) [inline docs/releases/2.0.txt]: (mention that it's only for newly-created fields)

File: tests/schema/tests.py [hunk 7]
```diff
@@ -520,6 +520,30 @@ def test_alter(self):
...
+        new_field.set_attributes_from_name("id")
+        new_field.model = Author
+        with connection.schema_editor() as editor:
+            editor.alter_field(Author, old_field, new_field, strict=True)
+
+    def test_alter_auto_field_to_char_field(self):
+        # Create the table
+        with connection.schema_editor() as editor:
+            editor.create_model(Author)
+        # Change AutoField to IntegerField
```
Comment (shaib) [inline tests/schema/tests.py]: to CharField

File: django/db/backends/oracle/operations.py [hunk 8]
```diff
@@ -29,12 +29,22 @@ class DatabaseOperations(BaseDatabaseOperations):
...
     seq_value integer;
+    seq_name user_tab_identity_cols.sequence_name%%TYPE;
 BEGIN
+    BEGIN
+        SELECT sequence_name INTO seq_name FROM user_tab_identity_cols
+        WHERE  table_name = '%(table_name)s' AND
+               column_name = '%(column_name)s';
+        EXCEPTION WHEN NO_DATA_FOUND THEN
+            seq_name := '%(no_autofield_sequence_name)s';
+    END;
```
Comment (felixxm) [inline django/db/backends/oracle/operations.py]: With extra `BEGIN/END` we can handle `NO_DATA_FOUND` exception (i.e. new-style sequence doesn't exist) and go forward with the old-style sequence name.

File: django/db/backends/oracle/operations.py [hunk 9]
```diff
@@ -29,12 +29,22 @@ class DatabaseOperations(BaseDatabaseOperations):
...
+
     SELECT NVL(MAX(%(column)s), 0) INTO table_value FROM %(table)s;
     SELECT NVL(last_number - cache_size, 0) INTO seq_value FROM user_sequences
-           WHERE sequence_name = '%(sequence)s';
+           WHERE sequence_name = seq_name;
     WHILE table_value > seq_value LOOP
-        seq_value := "%(sequence)s".nextval;
+        EXECUTE IMMEDIATE 'SELECT "'||seq_name||'".nextval FROM DUAL'
+        INTO seq_value;
```
Comment (felixxm) [inline django/db/backends/oracle/operations.py]: I know how it looks like, but there is no better way (in my opinion) to do this. We're establishing `seq_name` inside the trigger, therefore we cannot use bind variable for object name and we need to use string concatenation (see [doc](http://docs.oracle.com/database/121/LNPLS/dynamic.htm#LNPLS645)). I think that we can use
```SQL
seq_name := DBMS_ASSERT.ENQUOTE_NAME(seq_name, FALSE);
EXECUTE IMMEDIATE 'SELECT "'||seq_name||'".nextval FROM DUAL INTO seq_value;'
```
instead, but it seems to be overstatement for me.

Comment (shaib) [inline django/db/backends/oracle/operations.py]: I was referring to the part where we loop to get the sequence to advance beyond the largest value in the table.

File: django/db/backends/oracle/introspection.py [hunk 10]
```diff
@@ -76,14 +88,14 @@ def get_table_description(self, cursor, table_name):
         description = []
         for desc in cursor.description:
             name = force_text(desc[0])  # cx_Oracle always returns a 'str'
```
Comment (felixxm) [inline django/db/backends/oracle/introspection.py]: You're right. I created separate PR for this #8361.

File: django/db/backends/oracle/operations.py (hunk 9 above)
Comment (felixxm) [inline django/db/backends/oracle/operations.py]: We cannot set the current sequence value on Oracle. There are two ways to increase the current sequence value, the first one is to drop sequence and create new one with `START WITH` clause, the second one is to iterate sequence manually (like above).  On Oracle 12c we should be able to set sequence current value, but there is a huge [bug](https://emoracle.wordpress.com/2016/07/19/ora-00600-with-arguments-12811-154970/) in Oracle 12.1, therefore I decided to keep the simplest solution.

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: django/db/backends/base/schema.py [hunk 11]
```diff
@@ -966,3 +959,13 @@ def _constraint_names(self, model, column_names=None, unique=None,
...
+
+    def _delete_primary_key(self, model, strict=False):
+        constraint_names = self._constraint_names(model, primary_key=True)
+        if strict and len(constraint_names) != 1:
+            raise ValueError("Found wrong number (%s) of PK constraints for %s" % (
+                len(constraint_names),
+                model._meta.db_table,
+            ))
+        for constraint_name in constraint_names:
+            self.execute(self._delete_constraint_sql(self.sql_delete_pk, model, constraint_name))
```
Comment (felixxm) [inline django/db/backends/base/schema.py]: @apollo13 @timgraham Maybe I should send this change in a separate PR as a part of `_alter_field` cleanup (#8487).

Comment (apollo13) [inline django/db/backends/base/schema.py]: Whatever is easier for you (or makes more sense), I can rebase and fix conflicts as I go along…

Comment (timgraham) [inline django/db/backends/base/schema.py]: Splitting this into a few commits could make review easier. In particular the PR description has a few bullet points -- are those independent changes or all they all so tightly linked that they must be in the same commit?

Comment (felixxm) [inline django/db/backends/base/schema.py]: I can split these changes into three commits.

Comment (felixxm) [inline django/db/backends/base/schema.py]: Done.

**Comment by felixxm** [comment]
@shaib Thanks for review. I rebased from master.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: django/db/backends/oracle/operations.py [hunk 12]
```diff
@@ -528,13 +528,21 @@ def bulk_insert_sql(self, fields, placeholder_rows):
...
-                if not fields[i]:
-                    select.append(placeholder)
-                else:
+                if fields[i]:
                     internal_type = getattr(fields[i], 'target_field', fields[i]).get_internal_type()
-                    select.append(BulkInsertMapper.types.get(internal_type, '%s') % placeholder)
+                    placeholder = BulkInsertMapper.types.get(internal_type, '%s') % placeholder
+                # Add columns aliases to the first select to avoid "ORA-00918:
+                # column ambiguously defined" when two or more columns in the
+                # first select has the same values.
```
Comment (timgraham) [inline django/db/backends/oracle/operations.py]: has -> have
values -> value

File: django/db/backends/oracle/operations.py [hunk 13]
```diff
@@ -528,13 +528,21 @@ def bulk_insert_sql(self, fields, placeholder_rows):
...
+                if not query:
+                    placeholder = "%s col_%s" % (placeholder, i)
+                select.append(placeholder)
             query.append('SELECT %s FROM DUAL' % ', '.join(select))
-        return ' UNION ALL '.join(query)
+        # Bulk insert to the tables with Oracle identity columns causes that
```
Comment (timgraham) [inline django/db/backends/oracle/operations.py]: Bulk insert to tables with... (chop "the")

File: django/db/backends/oracle/operations.py [hunk 14]
```diff
@@ -528,13 +528,21 @@ def bulk_insert_sql(self, fields, placeholder_rows):
...
+        # Oracle internally adds sequence.nextval to it. Sequence.nextval
```
Comment (timgraham) [inline django/db/backends/oracle/operations.py]: causes Oracle to add

File: django/db/backends/oracle/operations.py [hunk 15]
```diff
@@ -528,13 +528,21 @@ def bulk_insert_sql(self, fields, placeholder_rows):
...
+        # cannot be used with the UNION operator. To prevent incorrect SQL move
```
Comment (timgraham) [inline django/db/backends/oracle/operations.py]: SQL,

File: django/db/backends/oracle/operations.py [hunk 16]
```diff
@@ -528,13 +528,21 @@ def bulk_insert_sql(self, fields, placeholder_rows):
...
+        # UNION to subquery.
```
Comment (timgraham) [inline django/db/backends/oracle/operations.py]: to a subquery

File: django/db/backends/oracle/operations.py [hunk 17]
```diff
@@ -528,13 +528,21 @@ def bulk_insert_sql(self, fields, placeholder_rows):
...
-                else:
+                if fields[i]:
                     internal_type = getattr(fields[i], 'target_field', fields[i]).get_internal_type()
-                    select.append(BulkInsertMapper.types.get(internal_type, '%s') % placeholder)
+                    placeholder = BulkInsertMapper.types.get(internal_type, '%s') % placeholder
+                # Add columns aliases to the first select to avoid "ORA-00918:
+                # column ambiguously defined" when two or more columns in the
+                # first select has the same values.
+                if not query:
+                    placeholder = "%s col_%s" % (placeholder, i)
```
Comment (timgraham) [inline django/db/backends/oracle/operations.py]: use single quotes

File: django/db/backends/oracle/introspection.py [hunk 18]
```diff
@@ -32,9 +35,14 @@ def get_field_type(self, data_type, description):
             precision, scale = description[4:6]
             if scale == 0:
                 if precision > 11:
-                    return 'BigIntegerField'
+                    if description.is_autofield:
+                        return 'BigAutoField'
```
Comment (timgraham) [inline django/db/backends/oracle/introspection.py]: I'd use: `return 'BigAutoField' if description.is_autofield else 'BigIntegerField'`

File: docs/releases/2.0.txt [hunk 19]
```diff
@@ -140,7 +140,8 @@ CSRF
 Database backends
 ~~~~~~~~~~~~~~~~~
 
-* ...
+* Oracle database introspection now works with auto fields, but only for newly
```
Comment (timgraham) [inline docs/releases/2.0.txt]: I think I'd put this under "Management commands", with wording like "On Oracle, inspectdb can now introspect AutoField, if the column is created as an identity column."

File: docs/releases/2.0.txt [hunk 20]
```diff
@@ -143,6 +143,10 @@ Database backends
 * Oracle database introspection now works with auto fields, but only for newly
   created.
 
+* Added support for the :class:`~django.db.models.AutoField`,
+  :class:`~django.db.models.BigAutoField` fields as Oracle Identity Columns
+  (:ticket:`26682`).
```
Comment (timgraham) [inline docs/releases/2.0.txt]: No ticket reference needed.

File: docs/releases/2.0.txt [hunk 21]
```diff
@@ -143,6 +143,10 @@ Database backends
 * Oracle database introspection now works with auto fields, but only for newly
   created.
 
+* Added support for the :class:`~django.db.models.AutoField`,
+  :class:`~django.db.models.BigAutoField` fields as Oracle Identity Columns
```
Comment (timgraham) [inline docs/releases/2.0.txt]: On Oracle, AutoField and BigAutoField are now created as identity columns. (explain the benefits of this or link "identity columns" to the Oracle docs? are there any concerns about having a mix of identity and non-identity columns in a project? should we provide a way to update existing columns to identity columns?)

File: django/db/backends/oracle/schema.py [hunk 22]
```diff
@@ -56,22 +63,35 @@ def alter_field(self, model, old_field, new_field, strict=False):
             # SQLite-ish workaround
             if 'ORA-22858' in description or 'ORA-22859' in description:
                 self._alter_field_type_workaround(model, old_field, new_field)
+            # If we're changing identity column type to non-numeric type
```
Comment (timgraham) [inline django/db/backends/oracle/schema.py]: If an identity column is changing to a non-numeric type, the drop the identity first.

File: django/db/backends/oracle/schema.py [hunk 23]
```diff
@@ -56,22 +63,35 @@ def alter_field(self, model, old_field, new_field, strict=False):
...
+            # we need to drop identity previous
+            elif 'ORA-30675' in description:
+                self._drop_identity(model._meta.db_table, old_field.column)
+                self.alter_field(model, old_field, new_field, strict)
+            # Modifying identity column isn't supported in Oracle. If we're
```
Comment (timgraham) [inline django/db/backends/oracle/schema.py]: If a primary key column is changing to an identity column, drop the primary key first.

File: django/db/backends/oracle/operations.py [hunk 24]
```diff
@@ -415,23 +396,32 @@ def sequence_reset_sql(self, style, model_list):
...
+                    table = self.quote_name(model._meta.db_table)
+                    column = self.quote_name(f.column)
+                    output.append(query % {
+                        'no_autofield_sequence_name': no_autofield_sequence_name,
+                        'table': table,
+                        'column': column,
+                        'table_name': strip_quotes(table),
+                        'column_name': strip_quotes(column),
+                    })
+
```
Comment (timgraham) [inline django/db/backends/oracle/operations.py]: chop blank line

File: django/db/backends/oracle/operations.py [hunk 25]
```diff
@@ -512,15 +502,26 @@ def combine_expression(self, connector, sub_expressions):
...
-        return truncate_name(trigger_name, name_length).upper()
+    def _get_sequence_name(self, cursor, table, pk_name):
+        cursor.execute("""
+            SELECT sequence_name
+            FROM user_tab_identity_cols
+            WHERE table_name = UPPER(%s)
+            AND column_name = UPPER(%s)""", [table, pk_name])
+        row = cursor.fetchone()
+        if row is None:
+            return self._get_no_autofield_sequence_name(table)
```
Comment (timgraham) [inline django/db/backends/oracle/operations.py]: `return ... if row is None else row[0]` ?

File: django/db/backends/oracle/schema.py [hunk 26]
```diff
@@ -31,10 +31,17 @@ def quote_value(self, value):
         else:
             return str(value)
 
+    def remove_field(self, model, field):
+        # If the column is the identity_column then drop identity is required
```
Comment (timgraham) [inline django/db/backends/oracle/schema.py]: If the column is an identity column, drop the identity before removing the field.

File: django/db/backends/oracle/schema.py [hunk 27]
```diff
@@ -56,22 +63,35 @@ def alter_field(self, model, old_field, new_field, strict=False):
...
 
     def _alter_field_type_workaround(self, model, old_field, new_field):
         """
         Oracle refuses to change from some type to other type.
         What we need to do instead is:
-        - Add a nullable version of the desired field with a temporary name
-        - Update the table to transfer values from old to new
-        - Drop old column
-        - Rename the new column and possibly drop the nullable property
+        - add a nullable version of the desired field with a temporary name,
```
Comment (timgraham) [inline django/db/backends/oracle/schema.py]: Removing upper case of each line and adding commas isn't required.

File: django/db/backends/oracle/schema.py [hunk 28]
```diff
@@ -56,22 +63,35 @@ def alter_field(self, model, old_field, new_field, strict=False):
...
+          if new column is a auto field then temporary column can not be
```
Comment (timgraham) [inline django/db/backends/oracle/schema.py]: . If the new column is an auto field, then the temporary column can't be nullable.

**Comment by felixxm** [comment]
@timgraham Thanks for review. I updated PR and reorganized commits.

File: docs/releases/2.0.txt (hunk 21 above)
Comment (felixxm) [inline docs/releases/2.0.txt]: I will try to answer these questions tomorrow.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: docs/releases/2.0.txt (hunk 21 above)
Comment (felixxm) [inline docs/releases/2.0.txt]: - are there any concerns about having a mix of identity and non-identity columns in a project?
> I don't think so.
-  should we provide a way to update existing columns to identity columns?
> There is no way to update existing columns to identity columns. Identity cannot be "attached" to existing columns. Theoretical scenario: drop PK constraint, add new identity column, copy data from old PK, adjust sequence value, drop old PK. But this scenario won't work for columns referenced by FKs (*ORA-02273: this unique/primary key is referenced by some foreign keys*), hence it's almost always useless. 
- explain the benefits of this or link "identity columns" to the Oracle docs?
> The biggest benefit is performance increase (see [article](https://oracle-base.com/articles/12c/identity-columns-in-oracle-12cr1)). The Oracle docs on this subject is laconic, but we can add a link to the official docs http://docs.oracle.com/database/121/DRDAA/migr_tools_feat.htm#DRDAA109.

**Comment by felixxm** [comment]
- are there any concerns about having a mix of identity and non-identity columns in a project?
> I don't think so.
-  should we provide a way to update existing columns to identity columns?
> There is no way to update existing columns to identity columns. Identity cannot be "attached" to existing columns. Theoretical scenario: drop PK constraint, add new identity column, copy data from old PK, adjust sequence value, drop old PK. But this scenario won't work for columns referenced by FKs (*ORA-02273: this unique/primary key is referenced by some foreign keys*), hence it's almost always useless. 
- explain the benefits of this or link "identity columns" to the Oracle docs?
> The biggest benefit is performance increase (see [article](https://oracle-base.com/articles/12c/identity-columns-in-oracle-12cr1)). The Oracle docs on this subject is laconic, but we can add a link to the official docs http://docs.oracle.com/database/121/DRDAA/migr_tools_feat.htm#DRDAA109.

File: django/db/backends/oracle/schema.py [hunk 29]
```diff
@@ -56,22 +63,34 @@ def alter_field(self, model, old_field, new_field, strict=False):
             # SQLite-ish workaround
             if 'ORA-22858' in description or 'ORA-22859' in description:
                 self._alter_field_type_workaround(model, old_field, new_field)
+            # If an identity column is changing to a non-numeric type, the drop
```
Comment (timgraham) [inline django/db/backends/oracle/schema.py]: chop "the"

File: docs/releases/2.0.txt [hunk 30]
```diff
@@ -140,7 +140,8 @@ CSRF
 Database backends
 ~~~~~~~~~~~~~~~~~
 
-* ...
+* On Oracle, ``AutoField`` and ``BigAutoField`` are now created as identity
```
Comment (timgraham) [inline docs/releases/2.0.txt]: I think this goes in "Models". Some stuff in "Database backends" in the 1.11 notes  should actually be in "Models."

File: tests/schema/tests.py [hunk 31]
```diff
@@ -532,6 +532,30 @@ def test_alter(self):
         self.assertEqual(columns['name'][0], "TextField")
         self.assertEqual(bool(columns['name'][1][6]), bool(connection.features.interprets_empty_strings_as_nulls))
 
+    def test_alter_auto_field_to_integer_field(self):
+        # Create the table
+        with connection.schema_editor() as editor:
+            editor.create_model(Author)
+        # Change AutoField to IntegerField
+        old_field = Author._meta.get_field("id")
```
Comment (timgraham) [inline tests/schema/tests.py]: Could use single quotes.

File: docs/releases/2.0.txt [hunk 32]
```diff
@@ -194,6 +194,8 @@ Management Commands
 
 * The new :option:`diffsettings --output` option allows formatting the output
   in a unified diff format.
+* On Oracle, inspectdb can now introspect ``AutoField``, if the column is
```
Comment (timgraham) [inline docs/releases/2.0.txt]: add backticks around inspectdb and add a blank line above.
chop the comme before "if"

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: django/db/backends/oracle/schema.py [hunk 33]
```diff
@@ -56,22 +63,34 @@ def alter_field(self, model, old_field, new_field, strict=False):
...
+          nullable.
         - Update the table to transfer values from old to new
         - Drop old column
         - Rename the new column and possibly drop the nullable property
         """
         # Make a new field that's like the new one but with a temporary
         # column name.
         new_temp_field = copy.deepcopy(new_field)
-        new_temp_field.null = True
+        new_temp_field.null = (new_field.get_internal_type() != 'AutoField')
```
Comment (shaib) [inline django/db/backends/oracle/schema.py]: Missed BigAutoField here, I think

File: django/db/backends/oracle/introspection.py [hunk 34]
```diff
@@ -7,6 +8,8 @@
 )
 from django.utils.deprecation import RemovedInDjango21Warning
 
+FieldInfo = namedtuple('FieldInfo', FieldInfo._fields + ('is_autofield',))
```
Comment (shaib) [inline django/db/backends/oracle/introspection.py]: I think I mentioned this before -- I find this name rebinding a bit odd. I'd like it better if the `FieldInfo` from `base.introspection` was imported as `BaseFieldInfo`.

File: django/db/backends/oracle/schema.py (hunk 33 above)
Comment (felixxm) [inline django/db/backends/oracle/schema.py]: Good catch! Thanks, I fixed this and added a new test: *schema.tests.SchemaTests.test_alter_int_pk_to_bigautofield_pk*.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
@shaib Thanks for review. `isort` failure is not related with this PR.

**Comment by felixxm** [comment]
Merged in 924a89e135fe54bc7622aa6f03405211e75c06e9 and 9af6c97504ef2b06dd5292d03ea94d3eb5d8c4d6.
